### ISSUE TITLE
mxPath: Adding undirected effect removes directed effect

### ISSUE THREAD

**Issue by jhorzek** [OpenMx/OpenMx#375]
Hi,
I've been experimenting with the `mxPath()` function lately, and have come across a problem that is probably an edge case. Adding both a directed effect and an undirected effect for the same observed variable does not seem to work. The effect added first is removed from the model.

```r
library(OpenMx)
model <- mxModel(type = "RAM",
                 manifestVars = c("x1", "x2"),
                 mxPath(from = "x1", to = "x2", 
                        arrows = 1, 
                        labels = "b_x2_x1"),
                 mxPath(from = "x1", 
                        to = "x2", 
                        arrows = 2, 
                        labels = "v_x2_x1"))
# adding the covariance removes the directed effect x1 -> x2
coef(model)
# v_x2_x1 
#      0 
model <- mxModel(type = "RAM",
                 manifestVars = c("x1", "x2"),
                 mxPath(from = "x1", 
                        to = "x2", 
                        arrows = 2, 
                        labels = "v_x2_x1"),
                 mxPath(from = "x1", to = "x2", 
                        arrows = 1, 
                        labels = "b_x2_x1"))
# adding the directed effect removes the covariance x1 <-> x2
coef(model)
# b_x2_x1
#      0 
```

Best,
Jannik

**Comment by mcneale** [comment]
That is possibly by design, although the paths would not overwrite each other (one goes in A and the other in the S matrix), so I think the restriction should be removed (other dev team members may disagree so we will discuss).  If you need a workaround, then you can add latent variables that cause the observed variables with paths equal to 1, give them variances of 1, and estimate their covariance instead while retaining the causal path from x1 to x2.

**Comment by jhorzek** [comment]
Thanks for the quick reply! I also think that adding both types of effects on the same variables is a rare situation and most likely a user error. However, we wanted to implement a model from [Gische & Voelkle, 2022](https://link.springer.com/article/10.1007/s11336-021-09811-z/figures/2) where there are both directed and undirected effects between repeatedly measured variables. As the autoregressive and cross-lagged effects in this model are constrained to be equal over time, the model should also allow the estimation of covariances.

**Comment by RMKirkpatrick** [comment]
Here's another possible workaround: directly modify the 'A' or 'S' MxMatrices.  For example, if you already have the undirected path in your model, you can add the directed one like this:
```
model$A$free["x2","x1"] <- TRUE
model$A$labels["x2","x1"] <- "b_x2_x1"
```
You can likewise modify the 'values' matrix of 'S' or 'A' to set start values.  Also, remember that 'S' is **s**ymmetric, so if you modify an off-diagonal element of 'S', you have to likewise modify the corresponding element on the other side of the diagonal.

**Comment by jhorzek** [comment]
Thank you! The workarounds work perfectly!

**Comment by tbates** [comment]
If you discuss this, plus 1 for removing the restriction in the next release: It surprises users when OpenMx silently deletes paths from their model, and as @mcneale notes, having both `A` and `S` paths connecting a pair of variables is a legitimate and valid model.

> That is possibly by design, although the paths would not overwrite each other (one goes in A and the other in the S matrix), so I think the restriction should be removed (other dev team members may disagree so we will discuss)

**Comment by mcneale** [comment]
Agreed.  It is not consistent with OpenMx philosophy: WYSIWWD - what you say is what we do.  Let's change it.

**Comment by RMKirkpatrick** [comment]
> Agreed. It is not consistent with OpenMx philosophy: WYSIWWD - what you say is what we do. Let's change it.

Be that as it may, this issue has straightforward workarounds, so I don't consider it that high-priority.

**Comment by RMKirkpatrick** [comment]
This is the change:
```
diff --git a/R/MxRAMModel.R b/R/MxRAMModel.R
index 676de747..928c317f 100755
--- a/R/MxRAMModel.R
+++ b/R/MxRAMModel.R
@@ -679,12 +679,6 @@ insertPathRAM <- function(path, model) {
                        A_labels[to, from] <- nextlabel
                        A_ubound[to, from] <- nextubound
                        A_lbound[to, from] <- nextlbound
-                       S_values[to, from] <- 0
-                       S_labels[to, from] <- as.character(NA)
-                       S_free[to, from] <- FALSE
-                       S_values[from, to] <- 0
-                       S_labels[from, to] <- as.character(NA)
-                       S_free[from, to] <- FALSE
                } else if (arrows == 2) {
                        S_free[to, from] <- nextfree
                        S_values[to, from] <- nextvalue
@@ -696,12 +690,6 @@ insertPathRAM <- function(path, model) {
                        S_labels[from, to] <- nextlabel
                        S_ubound[from, to] <- nextubound
                        S_lbound[from, to] <- nextlbound
-                       A_values[to, from] <- 0
-                       A_labels[to, from] <- as.character(NA)
-                       A_free[to, from] <- FALSE
-                       A_values[to, from] <- 0
-                       A_labels[to, from] <- as.character(NA)
-                       A_free[to, from] <- FALSE
                } else {
                        stop(paste("Unknown arrow type", arrows,
                                   "with source", omxQuotes(from),
```
It enables the requested behavior, and does not break any tests in 'make test'.  However, we should proceed carefully: we don't know how many careless users out there have scripts that implicitly rely on the old behavior to work properly.  There are backwards-compatibility considerations we should discuss.

**Comment by tbates** [comment]
Great job @RMKirkpatrick ! And less code :-)

The change would only impact models which included an unwanted path in error and then got lucky that it was accidentally removed via this un-stated rule :-) Seems a small (perhaps empty?) set of models, and they should not have worked/will not be working as coded.

On the upside, more people who assumed OpenMX would be WYSIWYG and didn't notice it was not in this one case will have their models fixed for them.

**Comment by RMKirkpatrick** [comment]
> The change would only impact models which included an unwanted path in error and then got lucky that it was accidentally removed via this un-stated rule :-) Seems a small (perhaps empty?) set of models, and they should not have worked/will not be working as coded.

Good point.  I just wanted the Team to reach consensus before we break backwards compatibility.

@jhorzek , would you mind if we adapted the example in your OP into a regression test for our test suite?  We would acknowledge you as a contributor to the package.

**Comment by jhorzek** [comment]
Hi @RMKirkpatrick, thank you very much for solving this issue! Please feel free to use the script as test. There is no need to add me as a contributor for this small code snippet, but thank you very much for the offer!

**Comment by RMKirkpatrick** [comment]
> Hi @RMKirkpatrick, thank you very much for solving this issue! Please feel free to use the script as test. There is no need to add me as a contributor for this small code snippet, but thank you very much for the offer!

OK, great.  I expect I will push the change to the master branch on Friday, after discussing it during the OpenMx Team meeting.

**Comment by mhunter1** [comment]
The OP's desired behavior is what LISREL paths do.

```
library(OpenMx)
model <- mxModel(type = "LISREL",
                 manifestVars = list(endo=c("x1", "x2")),
		     latentVars=list(endo=c("lx1", "lx2")),
                 mxPath(from = "lx1", to = "lx2", 
                        arrows = 1, 
                        labels = "b_x2_x1"),
                 mxPath(from = "lx1", 
                        to = "lx2", 
                        arrows = 2, 
                        labels = "v_x2_x1"))
# adding the covariance removes the directed effect x1 -> x2
coef(model)
# v_x2_x1 
#      0 
model <- mxModel(type = "LISREL",
                 manifestVars = list(end=c("x1", "x2")),
		     latentVars=list(endo=c("lx1", "lx2")),
                 mxPath(from = "lx1", 
                        to = "lx2", 
                        arrows = 2, 
                        labels = "v_x2_x1"),
                 mxPath(from = "lx1", to = "lx2", 
                        arrows = 1, 
                        labels = "b_x2_x1"))
# adding the directed effect does NOT remove the covariance x1 <-> x2
coef(model)
#b_x2_x1 v_x2_x1 
#      0       0 
```

**Comment by RMKirkpatrick** [comment]
We discussed this issue during the OpenMx dev team meeting this morning.  We did _not_ reach consensus as I hoped we would.  Some of us were in favor of the proposed change, and others were opposed to it.  Those in favor pointed out that having both a one-headed and two-headed arrow between two variables is a legitimate use-case, and that the "automagic" removal of paths is a deviation from OpenMx's WYSIWID design principles that surprises some users.  Those opposed emphasized the potentially substantial extent of backward-compatibility issues the change could engender, pointed out that relatively straightforward workarounds already exist, and countered that the current behavior is precisely what users should expect according to the RAM specification.

We agreed to discuss the issue again at our next meeting.  I am pretty sure that any change relevant to this issue that we make will differ from what is proposed upthread.  OpenMx's default behavior is probably going to remain the same; any new behavior is likely to be accessed by providing a non-default value for a new argument to `mxPath()`.

**Comment by RMKirkpatrick** [comment]
> OpenMx's default behavior is probably going to remain the same; any new behavior is likely to be accessed by providing a non-default value for a new argument to `mxPath()`.

During the OpenMx Team meeting today, we discussed what this new argument should be called, but didn't reach a decision.  Anyone have any suggestions?

A major reason why we didn't reach a decision is that I want to fix the bugs that @mhunter1 has highlighted here in this thread before I add new functionality to `mxPath()`.

**Comment by tbates** [comment]
An alternative would be a default that can be switched. Then a user building models that need covariance and causal paths between variables could just set  'leavePathsAlone = TRUE'

A nice aid to users would be to print out "Polite warning: I just deleted a path you added earlier because it is not my style to permit single and double headed paths to the same variable"

I only discovered this because I had developers near by at the boulder workshop to ask, so such a warning would save a lot of time and error.

My guess is that if openmx is forced to retain this silent delete behavior, then work arounds like a new parameter to the mxPath will leave coding too fragile: users will instead switch from the normal openmx style  to add separate error-variance latents for models of this type. Seems a pity but if you have to know that openmx is removing paths silently, have to mentally keep track of when, and special case a path to avoid it (and hopefully not have it deleted later) it will be easier to avoid altogether.

Perhaps in a few more years...

**Comment by mcneale** [comment]
I think the add an argument to mxPath() is least destructive of prior scripts that rely on the auto-delete feature.  It was part of the original RAM specification that one could not have both - possibly due to thinking that it would never be identified (which is incorrect).  The two people to ask about that - Jack McArdle and Rod McDonald - have both passed so we may never know their rationale.  Unless it's written down somewhere.  I personally feel that the autodelete is not consistent with OpenMx philosophy of what you say is what we do.  Possibly, a thorough rewrite in terms of nodes and edges will eliminate this historical accident.  Inertia says it'll take a good 10 years before it will, but maybe AI will speed up the inertia lag.

**Comment by RMKirkpatrick** [comment]
During the meeting this morning, we decided against the new argument to `mxPath()`.  For details, see my commit message for 70eeaff.
